### PR TITLE
Migrate BeneficiaryDetailFragment from XML to compose

### LINKED ISSUE #2593: refactor: Migrate BeneficiaryDetailFragment from XML to compose

### PR DESCRIPTION
**Pull request by itsPronay**
fixes: #2593  - Migrate BeneficiaryDetailFragment from XML to compose

before | after
----------|----------
image | image

[compose screen recording](https://github.com/openMF/mifos-mobile/assets/78101731/3e0fb392-f1ba-4117-aed0-f1ff303d693f)

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import org.mifos.mobile.R
+import org.mifos.mobile.core.ui.component.MifosTitleDescSingleLineEqual
+import org.mifos.mobile.core.ui.theme.MifosMobileTheme
+import org.mifos.mobile.models.beneficiary.Beneficiary
+
+@Composable
+fun BeneficiaryDetailContent(
+    beneficiary: Beneficiary?
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt:24]: beneficiary should not be null in here, 
check in the caller itself, if its null, show error ui state

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(16.dp)
+    ) {
+
+        MifosTitleDescSingleLineEqual(
+            modifier = Modifier
+                .padding(horizontal = 4.dp)
+                .padding(3.dp),
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt]: use dp in multiple of 4, so 4.dp

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            title = stringResource(id = R.string.beneficiary_name),
+            description = beneficiary?.name.toString()
+        )
+
+        MifosTitleDescSingleLineEqual(
+            modifier = Modifier
+                .padding(horizontal = 4.dp)
+                .padding(3.dp),
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt]: same here

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            modifier = Modifier.fillMaxWidth(),
+            colors = CardDefaults.outlinedCardColors(
+                containerColor = MaterialTheme.colorScheme.background
+            )
+        ) {
+            Column(
+                modifier = Modifier.padding(8.dp)
+            ) {
+                MifosTitleDescSingleLineEqual(
+                    modifier = Modifier.padding(vertical = 3.dp),
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt]: here also

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    title = stringResource(id = R.string.client_name),
+                    beneficiary?.clientName.toString()
+                )
+
+                MifosTitleDescSingleLineEqual(
+                    modifier = Modifier.padding(vertical = 3.dp),
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt]: ..

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                    title = stringResource(id = R.string.account_type),
+                    description = beneficiary?.accountType.toString()
+                )
+
+                MifosTitleDescSingleLineEqual(
+                    modifier = Modifier.padding(vertical = 3.dp),
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt]: ..

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                    title = stringResource(id = R.string.transfer_limit),
+                    description = beneficiary?.transferLimit.toString()
+                )
+
+                MifosTitleDescSingleLineEqual(
+                    modifier = Modifier.padding(vertical = 3.dp),
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailContent.kt]: ..

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+
+        return ComposeView(requireContext()).apply {
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+            setContent {
+                MifosMobileTheme {
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt]: use mifosComposeView

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                    BeneficiaryDetailScreen(
+                        beneficiary = beneficiary,
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt]: put beneficiary in viewmodel, and use expose directly from viewmodel to screen.

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+                        }
+                    }
+                },
+                colors = TopAppBarDefaults.topAppBarColors(
+                    containerColor = if (isSystemInDarkTheme())
+                        Color(0xFF1B1B1F)
+                    else
+                        Color(0xFFFEFBFF)
+                )
+            )
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: make another compose function

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+                }
+
+                is BeneficiaryUiState.ShowError -> {
+                    if (deleteBeneficiaryClicked) {
+                        Toast.makeText(context, uiState.message, Toast.LENGTH_SHORT).show()
+                        deleteBeneficiaryClicked = false
+                    }
+                }
+
+                else -> throw IllegalStateException("Undesired $uiState")
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: dont throw anything

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+    MifosMobileTheme {
+        BeneficiaryDetailScreen(
+            beneficiary = beneficiary,
+            uiState = BeneficiaryUiState.ShowError(R.string.error_deleting_beneficiary),
+            navigateBack = {},
+            updateBeneficiaryClicked = {},
+            deleteBeneficiary = {}
+        )
+    }
+}
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: use preview parameters, you can take reference from some screeens or my latest PR

**Review by AvneetSingh2001 — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+import org.mifos.mobile.ui.help.HelpViewModel
+import org.mifos.mobile.utils.Constants
+import org.mifos.mobile.utils.Network
+import org.mifos.mobile.utils.Toaster
+import org.mifos.mobile.viewModels.BeneficiaryDetailViewModel
+
+@AndroidEntryPoint
+class BeneficiaryDetailFragment : BaseFragment() {
+
+    private var beneficiary: Beneficiary? = null
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt]: remove this line, directly store the argument in the viewmodel, and only read from there

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    private val viewModel: BeneficiaryDetailViewModel by viewModels()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        arguments?.let {
+            beneficiary = it.getParcelable(Constants.BENEFICIARY)
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt]: directly pass to viewmodel

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
+    ): View {
+        return mifosComposeView(requireContext()) {
+            BeneficiaryDetailScreen(navigateBack = {
+                requireActivity().onBackPressedDispatcher.onBackPressed()
+            }, updateBeneficiaryClicked = {
+                updateBeneficiaryClicked()
+            })
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailFragment.kt]: indentation properly.

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+                viewModel.deleteBeneficiary(it)
+            })
+    } else {
+        BeneficiaryDetailScreen(beneficiary = beneficiary,
+            uiState = uiState,
+            navigateBack = navigateBack,
+            updateBeneficiaryClicked = updateBeneficiaryClicked,
+            deleteBeneficiary = {
+                viewModel.deleteBeneficiary(it)
+            })
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: dont do like this.

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+    Scaffold(topBar = {
+        BeneficiaryDetailTopAppBar(navigateBack = navigateBack,
+            updateBeneficiaryClicked = updateBeneficiaryClicked,
+            updateDropdownValue = { value ->
+                openDropdown = value
+                openDropdown
+            },
+            showAlert = {
+                showAlert = true
+            })
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: ) should be in next line

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+    var openDropdown by rememberSaveable { mutableStateOf(false) }
+    var showAlert by rememberSaveable { mutableStateOf(false) }
+    val context = LocalContext.current
+    var deleteButtonClicked by remember { mutableStateOf(false) }
+    var showBeneficiaryNullError by rememberSaveable {
+        mutableStateOf(true)
+    }
+
+    if (showAlert) {
+        MifosAlertDialog(onDismissRequest = {
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: onDismissRequest keep on next line

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+
+            when (uiState) {
+                BeneficiaryUiState.Initial -> Unit
+
+                BeneficiaryUiState.Loading -> {
+                    MifosProgressIndicator(
+                        modifier = Modifier
+                            .fillMaxSize()
+                            .background(MaterialTheme.colorScheme.background.copy(0.8f))
+                    )
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: use MifosProgressIndicatorOverlay

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+                        Toast.makeText(context, uiState.message, Toast.LENGTH_SHORT).show()
+                        deleteButtonClicked = false
+                    }
+                    if (showBeneficiaryNullError) {
+                        Toast.makeText(context, uiState.message, Toast.LENGTH_SHORT).show()
+                        showBeneficiaryNullError = false
+                    }
+                }
+
+                else -> Log.d("Error uiState", "Undesired $uiState\")")
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: remove else, we only have 4 states , so no need of else.

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt (hunk 16 above)
Comment (AvneetSingh2001) [reply app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: process in the viewmodel itself.

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailViewModel.kt [hunk 21]
```diff
@@ -1,21 +1,27 @@
...
 import javax.inject.Inject
 
 @HiltViewModel
-class BeneficiaryDetailViewModel @Inject constructor(private val beneficiaryRepositoryImp: BeneficiaryRepository) :
-    ViewModel() {
+class BeneficiaryDetailViewModel @Inject constructor(
+    private val beneficiaryRepositoryImp: BeneficiaryRepository
+) : ViewModel() {
 
+    var beneficiary: Beneficiary? = null
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailViewModel.kt]: use stateflow and merge with beneficiaryUiState

**Review by AvneetSingh2001 — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt (hunk 20 above)
Comment (itsPronay) [reply app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: `BeneficiaryUiState` has more than 4 states. Removing `else` will give us an error on `when(uiState)`

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+import org.mifos.mobile.viewModels.BeneficiaryDetailViewModel
+
+@Composable
+fun BeneficiaryDetailScreen(
+    viewModel: BeneficiaryDetailViewModel = hiltViewModel(),
+    navigateBack: () -> Unit,
+    updateBeneficiary: () -> Unit,
+) {
+    val uiState by viewModel.beneficiaryUiState.collectAsStateWithLifecycle()
+    val beneficiary : Beneficiary? = viewModel.getBeneficiary()
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: use collectAsStateWithLifecycle

**Review by AvneetSingh2001 — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+                        Toast.makeText(context, uiState.message, Toast.LENGTH_SHORT).show()
+                        deleteButtonClicked = false
+                    }
+                    if (showBeneficiaryNullError) {
+                        Toast.makeText(context, uiState.message, Toast.LENGTH_SHORT).show()
+                        showBeneficiaryNullError = false
+                    }
+                }
+
+                else -> Unit
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/ui/beneficiary_detail/BeneficiaryDetailScreen.kt]: we dont have to use else. Make a new UIState class specific to only this screen with above 4 subclasses. Then we can eliminate else case.

**Review by AvneetSingh2001 — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/utils/BeneficiaryDetailsUiState.kt [hunk 24]
```diff
@@ -0,0 +1,9 @@
+package org.mifos.mobile.utils
+
+sealed class BeneficiaryDetailsUiState {
+    object Initial : BeneficiaryDetailsUiState()
+    object Loading : BeneficiaryDetailsUiState()
+    object DeletedSuccessfully : BeneficiaryDetailsUiState()
+    data class ShowError(val message: Int) : BeneficiaryDetailsUiState()
+}
```
Comment (AvneetSingh2001) [inline app/src/main/java/org/mifos/mobile/utils/BeneficiaryDetailsUiState.kt]: move this in viewmodel

**Pull request merged**